_This project was developed independently as part of Code Institute’s Predictive Analytics Project. Any datasets or templates used are openly provided by the course or via public sources like Kaggle. All commentary and code logic are my own._

# Notebook 01: Data Collection

## Objectives
The purpose of this notebook is to gather, load, and perform a preliminary inspection of the raw data for the Heritage Housing project. This step ensures we're working with valid data sources, understand their structure, and prepare for cleaning and wrangling in the next phase.


### Inputs  
- Kaggle API token (`kaggle.json`)  
- Dataset from Kaggle: Ames Housing (Code Institute)

### Outputs  
- `/data/raw/house_prices_records.csv`  
- `/data/raw/inherited_houses.csv`  
- `/data/raw/house-metadata.txt`


### Change Working Directory

- To allow smooth access to the data files, we need to adjust our working directory. 
- Since this notebook lives in a subfolder (e.g. jupyter_notebooks), we need to change the working directory from its current folder to its parent folder

In [1]:
import os
# Smart Working Directory Setup 
project_root = '/workspaces/heritage_housing'
if os.getcwd() != project_root:
    try:
        os.chdir(project_root)
        print(f"[INFO] Changed working directory to project root: {os.getcwd()}")
    except FileNotFoundError:
        raise FileNotFoundError(f"[ERROR] Project root '{project_root}' not found!")

[INFO] Changed working directory to project root: /workspaces/heritage_housing


### Fetch Dataset from Kaggle

- To keep the workflow reproducible and professional, we will use Kaggle’s API to programmatically download the dataset.

Setup: Install Kaggle and Authenticate

In [2]:
# Install Kaggle CLI (if not already installed)
!pip3 install kaggle==1.5.12


[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip


- You must first create a Kaggle account and generate an API token from your account settings. This will download a kaggle.json file.
- Move kaggle.json to the root directory of this repo.
- Run the below to register the token and adjust permissions:

In [3]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = os.getcwd()
!chmod 600 kaggle.json  # Secure the file

### Download the Dataset

- We now fetch the dataset using the CLI. This project uses the "Heritage Housing Predictor" dataset from Kaggle competitions.

In [4]:
KaggleDatasetPath = "codeinstitute/housing-prices-data"
DestinationFolder = "data/raw"

!kaggle {KaggleDatasetPath} download -c heritage-housing-predictor -p {DestinationFolder}

usage: kaggle [-h] [-v] {competitions,c,datasets,d,kernels,k,config} ...
kaggle: error: argument command: invalid choice: 'codeinstitute/housing-prices-data' (choose from 'competitions', 'c', 'datasets', 'd', 'kernels', 'k', 'config')


- Unzip the Downloaded File

In [5]:
import zipfile
from pathlib import Path

for zip_file in Path(DestinationFolder).glob("*.zip"):
    with zipfile.ZipFile(zip_file, 'r') as z:
        z.extractall(DestinationFolder)
    zip_file.unlink()  # Delete zip file after extraction

### Load Required Libraries

In [6]:
# Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Load Datasets

In [7]:
# To read dataset from the csv file in a Pandas dataframe
df = pd.read_csv(f'data/raw/house_prices_records.csv')

# Load house sales records
house_prices_df = pd.read_csv('data/raw/house_prices_records.csv')

# Load inherited houses data
inherited_df = pd.read_csv('data/raw/inherited_houses.csv')

### Quick Peek at Each Dataset
- This gives us a rough idea of the data shape and the kind of features we’ll be dealing with.

In [8]:
print("\n--- House Prices Data ---")
print(house_prices_df.shape)
print(house_prices_df.columns)
house_prices_df.head()

print("\n--- Inherited Houses Data ---")
print(inherited_df.shape)
print(inherited_df.columns)
inherited_df.head()


--- House Prices Data ---
(1460, 24)
Index(['1stFlrSF', '2ndFlrSF', 'BedroomAbvGr', 'BsmtExposure', 'BsmtFinSF1',
       'BsmtFinType1', 'BsmtUnfSF', 'EnclosedPorch', 'GarageArea',
       'GarageFinish', 'GarageYrBlt', 'GrLivArea', 'KitchenQual', 'LotArea',
       'LotFrontage', 'MasVnrArea', 'OpenPorchSF', 'OverallCond',
       'OverallQual', 'TotalBsmtSF', 'WoodDeckSF', 'YearBuilt', 'YearRemodAdd',
       'SalePrice'],
      dtype='object')

--- Inherited Houses Data ---
(4, 23)
Index(['1stFlrSF', '2ndFlrSF', 'BedroomAbvGr', 'BsmtExposure', 'BsmtFinSF1',
       'BsmtFinType1', 'BsmtUnfSF', 'EnclosedPorch', 'GarageArea',
       'GarageFinish', 'GarageYrBlt', 'GrLivArea', 'KitchenQual', 'LotArea',
       'LotFrontage', 'MasVnrArea', 'OpenPorchSF', 'OverallCond',
       'OverallQual', 'TotalBsmtSF', 'WoodDeckSF', 'YearBuilt',
       'YearRemodAdd'],
      dtype='object')


1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       896         0             2           No       468.0          Rec   
1      1329         0             3           No       923.0          ALQ   
2       928       701             3           No       791.0          GLQ   
3       926       678             3           No       602.0          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotArea  \
0      270.0              0       730.0          Unf  ...    11622   
1      406.0              0       312.0          Unf  ...    14267   
2      137.0              0       482.0          Fin  ...    13830   
3      324.0              0       470.0          Fin  ...     9978   

   LotFrontage MasVnrArea  OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  \
0         80.0        0.0            0            6            5        882.0   
1         81.0      108.0           36            6            6       1329.0   
2         74.0        0.0           34            5            5        928.0   
3         78.0       20.0           36            6            6        926.0   

   WoodDeckSF  YearBuilt  YearRemodAdd  
0         140       1961          1961  
1         393       1958          1958  
2         212       1997          1998  
3         360       1998          1998  

[4 rows x 23 columns]

### Dataframe Summary
- the .info() methos can now be called on the dataframe object to read the dataframe summary. The result below presents the output into it's own datadrame for readability purposes.


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       1460 non-null   int64  
 1   2ndFlrSF       1374 non-null   float64
 2   BedroomAbvGr   1361 non-null   float64
 3   BsmtExposure   1422 non-null   object 
 4   BsmtFinSF1     1460 non-null   int64  
 5   BsmtFinType1   1315 non-null   object 
 6   BsmtUnfSF      1460 non-null   int64  
 7   EnclosedPorch  136 non-null    float64
 8   GarageArea     1460 non-null   int64  
 9   GarageFinish   1225 non-null   object 
 10  GarageYrBlt    1379 non-null   float64
 11  GrLivArea      1460 non-null   int64  
 12  KitchenQual    1460 non-null   object 
 13  LotArea        1460 non-null   int64  
 14  LotFrontage    1201 non-null   float64
 15  MasVnrArea     1452 non-null   float64
 16  OpenPorchSF    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  OverallQ

## Our Initial Observations

- house_prices_records.csv contains 1,460 rows and 24 columns. The target variable SalePrice is included.

- inherited_houses.csv contains 4 rows and 23 columns — all the same features except for the missing SalePrice column.

- No explicit ID column is shared between the datasets, so merging isn’t directly possible.

- Features include both numeric (e.g., `LotArea`, `YearBuilt`) and categorical data (e.g., `BsmtExposure`, `KitchenQual`).

- Likely nulls in columns like `GarageYrBlt`, `LotFrontage`, and basement features.

- Column names match well between datasets — suggesting they're aligned in structure.

## Summary of Actions Completed

- Changed working directory to project root.

- Installed Kaggle CLI and configured authentication.

- Programmatically downloaded and extracted all raw data from Kaggle.

- Loaded raw datasets and confirmed structural consistency.

- Inspected column names, dimensions, and got a feel for data types and formats.

- Noted down initial findings to guide the cleaning strategy.

## Next Steps

Our acquired raw datasets are now prepared for:


### Preprocessing and Data Cleaning:

- Address missing values for columns such as `LotFrontage`, `GarageYrBlt`, and `BedroomAbvGr` in house_prices_records.csv.
- Examine and deal with columns like `EnclosedPorch` and `WoodDeckSF` that have a lot of null values to ascertain their applicability.
- To make integration and analysis easier, align the two datasets' layout and structure.
- Where there may be differences, standardize the types of columns (floats vs. integers, for example).

### EDA (Exploratory Data Analysis):

- To find important features, look into the relationships between house attributes and sale prices in house_prices_records.csv.
- Use visual aids, such as heatmaps or scatter plots, to help direct feature selection and model construction.
- Examine any connections that might exist between the data in the larger dataset and the characteristics of the inherited homes.

